### PR TITLE
Vignette

### LINKED ISSUE #3: Vignette to compare hypatia and squire outputs
We need an easy way to validate the implementation of hypatia.

We can create a vignette which plots both prevalence curves for 2 years in the conditions from #2 

>     Epidemics to be simulated for 1 fixed value of R0 over time
>     Epidemics to be simulated for any countries contained in squire::population
>     Epidemics assumed to have infinite healthcare capacity

We can compare for R0 = c(2, 3, 4) and countries = c('India', 'Nigeria', 'Panama')

### PR DESCRIPTION
**Pull request by ceeb950**
Added vignette but code is failing. create_setup_process in processes.R is failing on exposed[!hosp] expecting it to be 1 value.

### CODE REVIEW

**Comment by ceeb950** [comment]
@OJWatson  - please look review this. I have taken a lot from your dev/oj-refactor and both branches now fail for the same reason. Pull request: https://github.com/mrc-ide/hypatia/pull/38

**Comment by OJWatson** [comment]
The error is not the same as in dev/oj_refactor, which is passing - https://github.com/mrc-ide/hypatia/actions/runs/394994811.

Just spotted you have been doing lots of print statements to help debug. Couple of tricks that might help and are how I found what is causing the issue in the code as it is in the vignette branch. 

Firstly, I went to the `introduction.R` vignette to try recreate the error I think you are working through. (The `introduction.R` vignette didn't work as you have copied over from the introduction vignette on the dev/oj_refactor as it was but lots of functions in the main branch have changed and/or have now been set up differently.) I hit this error first which is not the one I think you sent me:

```
output <- run_simulation(pop, parameters)

 Error in individual::state_count_renderer_process(individuals$human$name,  : 
  Expecting a single string value: [type=NULL; extent=0]. 
```

So the first thing I did was run the function `traceback()`. This is the same as clicking on the Show Traceback button: 

image

This gave the following:

```
5. stop(structure(list(message = "Expecting a single string value: [type=NULL; extent=0].", 
    call = individual::state_count_renderer_process(individuals$human$name, 
        unlist(lapply(states, "[[", "name"))), cppstack = NULL), class = c("Rcpp::not_compatible", 
"C++Error", "error", "condition"))) 
4. individual::state_count_renderer_process(individuals$human$name, 
    unlist(lapply(states, "[[", "name"))) at processes.R#294
3. create_processes(states, variables, events, individuals) 
2. individual::simulate(individuals = individuals, processes = create_processes(states, 
    variables, events, individuals), end_timestep = parameters$time_period, 
    parameters = parameters, initialisation = create_setup_process(individuals, 
        states, events, variables)) at model.R#32
1. run_simulation(pop, parameters) 
```
Following through that took me to the `create_processess` function which is:

```
individual::state_count_renderer_process(
         #individuals$name,
         #statesnamevector
         individuals$human$name,
         unlist(lapply(states, "[[", "name"))
      )
```

I think you may have figured it out that you need to have `individuals$name`. The reason for this is because the `create_human` function creates individuals differently to how I set it up up in dev/oj_refactor. In this branch we have the following:

```
create_human <- function(
  states,
  variables,
  events) {

  human <- individual::Individual$new(
    "human",
    states = states,
    variables = variables,
    events = events
  )

  human
}
```

This is returning an `individual::Individual` object. In the dev/oj_refactor branch we were doing something different when creating our individuals:

```
create_individuals <- function(states,
                               variables,
                               events,
                               parameters) {
  human <- individual::Individual$new(
    "human",
    states = list(
      states$S,
      states$E,
      states$IMild,
      states$ICase,
      states$IOxGetLive,
      states$IOxGetDie,
      states$IOxNotGetLive,
      states$IOxNotGetDie,
      states$IMVGetLive,
      states$IMVGetDie,
      states$IMVNotGetLive,
      states$IMVNotGetDie,
      states$IRec,
      states$R,
      states$D
    ),
    
    variables = variables,
    
    events = events
  )
  
  list(human = human)
}
```

Here we are returning a list with the first element being a named `individual::Individual` object. So changing that in `create_processes` then led to the related error in the infection process:

image

So again going to the last hypatia process it breaks shows that we still have the `infection_process` thinking that it is expecting a list (i.e. getting humans by `individuals$human` when what you would then want is `indidivudals`), e.g.:

```
# Not this
# IMild <- api$get_state(individuals$human, states$IMild)
# This
IMild <- api$get_state(individuals, states$IMild)
```

After fixing those (and changing a couple of duff tests) I was able to get this to pass locally. I pushed these changes to `dev/vignette_fix`. I did notice that when i was testing this in Github though that it was failing as it was requesting `individual v0.0.8` and so specifying 0.0.9 has now passed - https://github.com/mrc-ide/hypatia/actions/runs/395018649.

@giovannic at some point I have clearly been using v0.0.9 which works but I have noticed that `indivdual` is only at 0.0.8 and there is a release for 0.0.9. If using v0.0.8 then you get the following errors shown here https://github.com/mrc-ide/hypatia/runs/1483144043?check_suite_focus=true#step:9:29:

```
Expecting a single value: [extent=19].
```

I **think** Carol that this was the error you were getting. So I think best to bring Giovanni in on this one as this is related to `individual` I think rather than `hypatia` or rather the way that we are using `individual` in `hypatia`.

**Comment by giovannic** [comment]
Agreed, it looks like the issue is we need individual:0.0.9 for the stochastic scheduling.

`Expecting a single value: [extent=19].` Means that the vignette has passed in a vector (of size 19) when individual was expecting 1 number. Coincidentally 0.0.9 introduces stochastic scheduling, so that several delays can be set instead of just one.

It doesn't look like there'll be any breaking changes, so try bumping the version.

**Comment by ceeb950** [comment]
@richfitz -  Rich you may want to review this - I am unable to get the comparisons on the Validation.html file to work so have temporarily commented that last section out. I added Giovanni's infection tests (#33) (test-infection_process.R) but they are failing and need other info and files to work. I have skipped them for now. So it would be appreciated if you can look at this for any glaring errors for the minute and I will continue to work on the comparisons and OJ's comment at the bottom of the Validation.Rd file.   https://github.com/mrc-ide/hypatia/pull/38

File: .github/workflows/R-CMD-check.yaml [hunk 1]
```diff
@@ -5,6 +5,7 @@ on:
     branches:
       - main
       - master
+      - dev/vignette_fix
```
Comment (richfitz) [inline .github/workflows/R-CMD-check.yaml]: this can come out (looks like from @OJWatson :grin: )

File: R/individuals.R [hunk 2]
```diff
@@ -22,3 +22,40 @@ create_human <- function(
   human
 }
 
+#' @title Define the human model returning more than 1 human
```
Comment (richfitz) [inline R/individuals.R]: this code looks unrelated to the vignette? was it included in error?

File: tests/testthat/test-events.R [hunk 3]
```diff
@@ -21,7 +23,7 @@ test_that("create_event_based_processes assigns a listener to each event", {
     states,
     variables,
     events,
-    parameters
+    psq
```
Comment (richfitz) [inline tests/testthat/test-events.R:25]: this change suggests that this test is incomplete and would have failed if it was used (i.e., the `parameters` promise was never forced). The expectation below seems very minimal; we might add a ticket to expand this later with some indication of what the test here is for

File: tests/testthat/test-infection_process.R [hunk 4]
```diff
@@ -0,0 +1,139 @@
+test_that("infection_process does not schedule infections when no-one is infected", {
+skip("skip test")
```
Comment (richfitz) [inline tests/testthat/test-infection_process.R]: Should these be skipped? There's not a lot of point of merging in a file of non-working skipped tests. Was it intended to be introduced in a PR that adds a vignette?

File: vignette/Validation.Rmd [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+  fig.path = "man/figures/README-",
+  fig.width = 9, 
+  fig.height = 9
+)
+```
+
+```{r setup}
+library(hypatia)
+library(ggplot2)
+library(tidyverse)
```
Comment (richfitz) [inline vignette/Validation.Rmd]: Your vignette must not use any packages that are not listed in the Suggests (or Imports) of the DESCRIPTION, and tidyverse is not a package that should be depended on by other packages. What do you need from it? Please replace with imports to the needed packages (looks like tidyr, tibble and dplyr)

Given that you namespace qualify the hypatia calls, you don't need `library(hypatia)` either.

File: vignette/Validation.Rmd [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+```
+
+```{r hypatia}
+
+# Create our population
+countryname = "Antigua and Barbuda"
+
```
Comment (richfitz) [inline vignette/Validation.Rmd]: Please use `<-` for assignment

File: vignette/Validation.Rmd [hunk 7]
```diff
@@ -0,0 +1,155 @@
...
+pop <- hypatia:::get_population(countryname)
```
Comment (richfitz) [inline vignette/Validation.Rmd]: This suggests that get_population should be part of the API as you're advertising it to  users.

File: vignette/Validation.Rmd [hunk 8]
```diff
@@ -0,0 +1,155 @@
...
+parameters <- get_parameters(
+  population = pop$n,
+  contact_matrix_set = squire::get_mixing_matrix(iso3c = "ATG"),
+  R0 = 2,
+  time_period = time_period,
+  tt_contact_matrix = 0
+)
+
+# run our simulation - default timesteps = 200
+df <- hypatia:::run_simulation(pop, parameters, max_age = 100)
```
Comment (richfitz) [inline vignette/Validation.Rmd]: This function is part of the API so you should use `::` not `:::`

File: vignette/Validation.Rmd [hunk 9]
```diff
@@ -0,0 +1,155 @@
...
+
+numruns <- length(df)
+numdatapoints <- length(df[[1]][[1]]) - 1
```
Comment (richfitz) [inline vignette/Validation.Rmd]: I think this is meant to be
```suggestion
numdatapoints <- nrow(df) - 1
```

however, as it is we get

```
df[[1]] # the column timestep
df[[1]][[1]] # the first value of that column
length(df[[1]][[1]]) # always 1 because of the second `[[` operation
```

which means the expression always evaluates to zero. However it is unused so can be removed.

File: vignette/Validation.Rmd [hunk 10]
```diff
@@ -0,0 +1,155 @@
...
+
+pop <- hypatia:::get_population(countryname)
+
+# Scale it for speed
+pop$n <- round(pop$n/10)
+
+# Create our equivalent parameters
+time_period <- 200
+
+parameters <- get_parameters(
```
Comment (richfitz) [inline vignette/Validation.Rmd]: unlike the other functions this one does not have `hypatia::` - either all calls into the package should be namespace qualified or all (I'd prefer all)

File: vignette/Validation.Rmd [hunk 11]
```diff
@@ -0,0 +1,155 @@
...
+                           "human_D_count")
+
+  # Convert to long format
+df <- tidyr::pivot_longer(tibble::as_tibble(df), all_of(statesnamevector))
+
+strname <- "Hypatia Model Simulation"
+subtitle <- paste(
+    "Simulation for 1 run and 1000 data points")
+
+time <- seq(1,200,1)
```
Comment (richfitz) [inline vignette/Validation.Rmd]:
```suggestion
time <- seq(1, time_period, 1)
```

File: vignette/Validation.Rmd [hunk 12]
```diff
@@ -0,0 +1,155 @@
...
+
+timsestep <- df$timestep
+value <- df$value
+name <- df$name
+group <- df$group
```
Comment (richfitz) [inline vignette/Validation.Rmd]: these four lines should be deleted as they are not needed (to illustrate, notice how `timsetep` is misspelt, but `timestep` is still found from `df` due to the way that ggplot works)

```suggestion
```

File: vignette/Validation.Rmd [hunk 13]
```diff
@@ -0,0 +1,155 @@
...
+  )
+
+plot(out)
+
+```
+
+Get the actual comparative outputs. 
+
+```{r comparisons}
+#
```
Comment (richfitz) [inline vignette/Validation.Rmd]: please either fix or remove

File: vignette/Validation.Rmd [hunk 14]
```diff
@@ -0,0 +1,155 @@
...
+# 
+# ggplot(rbind(sq, hyp), aes(t, y, color = model)) + 
+#   geom_line() + 
+#   facet_wrap(~compartment, scales = "free")
+
+
+
+```
+
+## License
```
Comment (richfitz) [inline vignette/Validation.Rmd]: no need to add a licence. comment below looks like it was included in error?

File: vignette/Validation.Rmd [hunk 15]
```diff
@@ -0,0 +1,155 @@
...
+    "Simulation for 1 run and 1000 data points")
+
+time <- seq(1,200,1)
+
+timsestep <- df$timestep
+value <- df$value
+name <- df$name
+group <- df$group
+
+ggplot2::ggplot(
```
Comment (richfitz) [inline vignette/Validation.Rmd]: Can you get rstudio to reindent this please? it should not line up against the left edge after the first `+`

File: vignette/Validation.Rmd [hunk 16]
```diff
@@ -0,0 +1,155 @@
+---
```
Comment (richfitz) [inline vignette/Validation.Rmd]: This vignette is in the wrong directory, which is why it is not being built on CI (that would have errored in R CMD check due to the unstated dependencies). It should be in `vignettes` (with an s) not `vignette`. Normally I do

```
usethis::use_vignette("vignettename")
```

to do the initial setup as it is error-prone to do by hand.

File: R/infection_process.R [hunk 17]
```diff
@@ -0,0 +1,53 @@
+#' @title Calculating the FOI
```
Comment (richfitz) [inline R/infection_process.R]: Looks like this is introduced in #33 so I presume this PR is queued behind that?

**Review by richfitz — CHANGES_REQUESTED** [review]
Looks like a couple of unrelated things here, with code changes and test changes unrelated to the PR in question. Most look like they come from #33 (though some look like they'll take some work to integrate as there will be conflicts) but others like `create_individuals` appear unrelated and are not used by any code in the package (removing this will fix the coverage failure). I've skipped over the package changes to re-look at that after the other PRs are through.

The vignettes are not being built on CI, due to the directory name being incorrect.

There are also files that should be removed:

* Remove vignette/Validation.html
* Remove vignette/man/figures/README*

A .gitignore rule of `*.html` will help with the Validation.html one but the others are more surprisingly included. *.png might be enough in .gitignore

File: tests/testthat/test-infection_process.R (hunk 4 above)
Comment (richfitz) [reply tests/testthat/test-infection_process.R]: Sorry, I see now that this is waiting until #33 is resolved?

File: .github/workflows/R-CMD-check.yaml (hunk 1 above)
Comment (ceeb950) [reply .github/workflows/R-CMD-check.yaml]: done

File: vignette/Validation.Rmd (hunk 12 above)
Comment (ceeb950) [reply vignette/Validation.Rmd]: I get warning when this is not done.

File: R/processes.R [hunk 18]
```diff
@@ -263,39 +263,28 @@ create_event_based_processes <- function(
...
-      statesnamevector
-   )
+    infection_process(individuals, states, variables, events),
 
-   processes <- list(
-      process_render
-   )
+    individual::state_count_renderer_process(
+      individuals$name,
+      unlist(lapply(states, "[[", "name"))
```
Comment (OJWatson) [inline R/processes.R:292]:
```suggestion
      vapply(states, "[[", FUN.VALUE = character(1), "name")
```

**Review by OJWatson — APPROVED** [review]

**Comment by OJWatson** [comment]
Overall looks fine to me. You need to address these merge conflicts but otherwise looks okay.

**Comment by ceeb950** [comment]
@OJWatson - for review

**Comment by ceeb950** [comment]
@richfitz - code now passes all tests - please review - https://github.com/mrc-ide/hypatia/pull/38

File: DESCRIPTION [hunk 19]
```diff
@@ -11,7 +11,7 @@ License: MIT + file LICENSE
 Encoding: UTF-8
 LazyData: true
 Remotes:  
-    mrc-ide/individual@0.0.8,
+    mrc-ide/individual@0.0.9,
     mrc-ide/squire@odin-1-1-0
```
Comment (richfitz) [inline DESCRIPTION:15]: The squire update has been merged now, so this can be reset

```suggestion
    mrc-ide/squire@v0.5.8
```

File: R/processes.R (hunk 18 above)
Comment (richfitz) [reply R/processes.R:292]: Is there a reason for this change? It looks empirically that is not needed, and I do not see where the `vcapply` version would fail where that would **not** be the desired outcome

**Review by richfitz — APPROVED** [review]
Thanks! Two small things here that can be mopped up later, but let's just merge this for now.
